### PR TITLE
Blog post about TLS logging improvements in OTP 22

### PR DESCRIPTION
**Pull request by peterdmv**
Blog post about TLS logging improvements in OTP 22.

### CODE REVIEW

**Comment by garazdawi** [comment]
Nice!

Some ideas:
* Maybe include how it was implemented using the new logger framework?
* Maybe use a common error example that the logging can help to find? Miss-configured ciphers or something like that?
* Comparison with openssl logging output? Just to make it clear that they are (almost?) identical.

**Comment by KennethL** [comment]
I suggest that you mention that we are working on TLS 1.3 support in the beginning, then it is more logical that you mention the tls1.3 tag at the end.

**Comment by IngelaAndin** [comment]
Maybe mention that is not necessarily the  final format, as the more Erlang like part of the log can be further improved to  make it even more useful, that is cipher suites can be converted to  the Erlang representation and make more sense to a human.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+Finally the logger can be called with the debug level and domain parameters.
+
+```erlang
+ssl_logger:debug(Opts#ssl_options.log_level,
+	         Report,
+		 #{domain => [otp,ssl,handshake]}),
+```
+
+Those who are interested of the current state of development can already play with the
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: interested in the

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        lt ->
+            ?LOG_NOTICE(Report);
+        eq ->
+            ?LOG_NOTICE(Report);
+        _ ->
+            ok
+    end.
+```
+
+Finally the logger can be called with the debug level and domain parameters.
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: Finally, the

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+```erlang
+Config = #{level => debug,
+           filter_default => stop,
+           formatter => {ssl_logger, #{}}},
+```
+
+Handler filter level is set to ```debug``` with ```stop``` as the default filter action. We also
+need a filter that lets the log events pass to the formatter if the source of the log event is the
+ssl application. In other words we need a domain filter with the action ```log``` on all sub-domains
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: In other words, we

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+[Truncated for brevity]
+```
+
+The verbose debug logging proved to be especially useful during the development of new
+extensions as previously we had to use wireshark captures to validate TLS protocol
+messages.
+
+## Implementation
+
+In the ```ssl``` application we needed a way to handle two types of protocol
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: In the ```ssl``` application, we

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+      <<0,50>>,
+      <<192,4>>,
+      <<192,14>>]},
+ {compression_methods,[0]},
+...
+[Truncated for brevity]
+```
+
+This is not the final format as there are many ways to further improve the representation
+of the handshake protocol messages such as converting the cipher suites to a human readable
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: human-readable

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+---
+Erlang/OTP 22 will be an important release for the ```ssl``` application. We are working on
+several new features and improvements such as support for TLS 1.3, some of those are already
+on the master branch. This blog post presents the new ssl debug logging built on the new
+logger API.
+
+## Usage
+
+As the ```ssl``` application undergoes a lot of changes the release of the new logger API
+presented the opportunity to level up its debug logging capabilities to be on pair with
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: on par with

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+OpenSSL.
+
+We introduced a new option ```log_level``` that specifies the log level for the ```ssl```
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: We have introduced

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+```
+
+The verbose debug logging proved to be especially useful during the development of new
+extensions as previously we had to use wireshark captures to validate TLS protocol
+messages.
+
+## Implementation
+
+In the ```ssl``` application we needed a way to handle two types of protocol
+messages (tls_record and handshake) each with a custom formatter function.
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: messages, tld_record and handshake, each

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+
+The most straightforward solution was to add a new handler instance to the logger with a
+special formatter function that filters out all the "noise" coming from other
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: with a custom formatter that

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+modules of the system.
+
+The handler itself could be based on ```logger_std_h```, there was no need to implement
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: I get what you want to say here, but I think I would reword it somewhat. I don't really have any good suggestions on what to write instead...

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+            BinMsg = lists:flatten(BinMsg0),
+            format_tls_record(Direction, BinMsg);
+        'handshake' ->
+            format_handshake(Direction, BinMsg0);
+        _Other ->
+            []
+    end.
+```
+
+There are two more helper functions that wrap around the logging macros:
```
Comment (garazdawi) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: Why do you have these helpers?

**Review by garazdawi — CHANGES_REQUESTED** [review]
Great additions, looks good.

A question that came to me is why ssl keeps its own debug level and it is not configured through the handler?

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 1 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 2 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 3 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 4 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 5 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 6 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 7 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 8 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 9 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 11 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: The main reason was to be able to set logging level per TLS session. I will update the text.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 10 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will reword.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 11 above)
Comment (garazdawi) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: aha!

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+Putting it all together we get the following function.
+
+```erlang
+start_logger() ->
+    Config = #{level => debug,
+               filter_default => stop,
+               formatter => {ssl_logger, #{}}},
+    Filter = {fun logger_filters:domain/2,{log,sub,[otp,ssl]}},
+    logger:add_handler(ssl_handler, logger_std_h, Config),
+    logger:add_handler_filter(ssl_handler, filter_non_ssl, Filter).
```
Comment (sirihansen) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: This is really a question about the implementation, not about the blog post: Why do you add the filter separately instead of including it directly in the `Config` map:
```
Config = #{level => debug,
￼                filter_default => stop,
                filters => [{filter_non_ssl,{fun logger_filters:domain/2,{log,sub,[otp,ssl]}}}],
￼                formatter => {ssl_logger, #{}}},
```

File: _posts/2018-10-03-ssl-logging-in-otp-22.md [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+        lt ->
+            ?LOG_NOTICE(Report);
+        eq ->
+            ?LOG_NOTICE(Report);
+        _ ->
+            ok
+    end.
+```
+
+Finally, the logger can be called with the debug level and domain parameters.
```
Comment (sirihansen) [inline _posts/2018-10-03-ssl-logging-in-otp-22.md]: I got a bit confused by this sentence - mostly since you say that "the logger can be called", while the example really calls ssl_logger. Maybe it would be clearer if you state something like "To print a log event, the above functions are called with the configured ssl log level and the domain parameter".

Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: Will fix.

File: _posts/2018-10-03-ssl-logging-in-otp-22.md (hunk 12 above)
Comment (peterdmv) [reply _posts/2018-10-03-ssl-logging-in-otp-22.md]: I didn't realize that it can be also configured this way. Thanks for the tip!
